enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,505 @@
+of the CoreDNS repository. As the plugin will inspect the DNS lookup queries and response from other plugins, it needs to be added before the other plugins
+(namely `forward` plugin) which takes care of the DNS lookups for the DNS names external to the cluster.
+
+The `egressfirewall` plugin will watch the `EgressFirewallDNSName` CRs and whenever there is a DNS lookup which matches one of the `EgressFirewallDNSName`
+CRs (either regular or wildcard DNS names or both), then it will update the `.status` of the `EgressFirewallDNSName` CR(s) if there's any change
+in the corresponding IP addresses and/or the next lookup time information (TTL + last lookup time). The process is explained in the [Workflow Description](#workflow-description) section.
+
+`SharedIndexInformer` will be used for tracking events related to `EgressFirewallDNSName` CRs. The details about the regular and wildcard DNS names will be stored
+in two separate maps by the plugin. Whenever there is a DNS lookup, if there is no `EgressFirewallDNSName` CRs created, then the `egressfirewall` plugin will just
+send the received response to the lookup. A DNS name will be checked for a match in both the maps. If there is no match then also the plugin will just send the
+received response to the lookup. However, when a match is found (in either of the maps or in both), then the corresponding `EgressFirewallDNSName` CR is updated
+with the current IP addresses along with the corresponding TTL and current time as the last lookup time, if the same information is not already available.
+
+#### OVN-K cluster manager
+
+For every unique DNS name used in EgressFirewall rules, OVN-K cluster manager will create a corresponding `EgressFirewallDNSName` CR.
+The name of the CR will be assigned using a hash function (similar to the ComputeHash
+[here](https://github.com/openshift/kubernetes/blob/master/pkg/controller/controller_utils.go#L1157-L1172))
+prefixed by `dns-`. It will also delete a `EgressFirewallDNSName` CR, when all the rules containing the corresponding DNS name are deleted.
+
+#### OVN-K master(s)
+
+The OVN-K master(s) will watch the `EgressFirewallDNSName` CRs. Whenever the `.status` of the CRs will be updated with new IP addresses and corresponding
+TTL information for a DNS name, OVN-K master(s) will update the `AddressSet` mapped to the DNS name. This `AddressSet` will be linked
+to the ACL rule(s) for the EgressFirewall rule(s) in which the DNS name is used. This will ensure that the latest IP addresses are always updated in
+the `AddressSets`.
+
+
+### Risks and Mitigations
+
+* The `EgressFirewallDNSName` CR will be created by OVN-K Cluster manager whenever a new DNS name is used in a EgressFirewall rule. The CR will be deleted
+when the corresponding DNS name is not used in any of the EgressFirewall rules. The `EgressFirewallDNSName` CR should not be modified (created or deleted or
+updated) by an user. Doing so may lead to undesired behavior of EgressFirewall.
+
+### Drawbacks
```

Comment:
maybe we can add a note here about extra delay for requests matching wildcarded DNS names, that don't have a wildcarded DNS record?